bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthStoreHandlerImpl.java

```diff
@@ -468,8 +424,6 @@ public void close() {
                 // re-setting thread state to interrupted
                 Thread.currentThread().interrupt();
             } finally {
-                // force to remove the service no matter we got the lock or not
-                this.storage = null;
```

Comment:
don't we want to release the storage reference once `close` was called?

Reply:
The caller of this `close()` method releases the `storageFactory` by [assigning](https://github.com/eclipse/smarthome/pull/6546/files#diff-7ffd2eb72c9e9984fa8fb836b9c98d13R205) it `null`. That will  release the object containing this `storage` variable and with that the storage I assume.

Reply:
Hm, this comment just had me have a deeper look into the code.
If we set it to null here, we will have to deal with null values for storage throughout this class (so those checks must not be removed either).
I guess this is the rea
The issue is that the consumers are NOT using OSGi services, so we cannot rely on the OSGi lifecycle management. If the storage disappears from the factory, the factory will go down, but the `OAuthClientService`s used by the consumers would be happily used and with them the (outdated) storage.
I guess this is the reason behind the comment I mentioned in https://github.com/eclipse/smarthome/pull/6546#pullrequestreview-177931953.
So I'd assume that we would still have to deal with a null storage after the factory has been deactivated.

Reply:
I'm not sure if I completely understand what you mean. But with the current suggested change if the storage would be unregistered the `StorageFactory` will be set to null. So if a consumer that wants to use it will be met with a null pointer exception because storageFactory is null. Not the nicest solution I guess. But setting a service to null seems to be a common thing to do. I've seen it everywhere. And I don't think I've seen code that deals with such situations. Which would mean to wrap every call to a service in a null check. I haven't enough knowledge of the system why this would be different for the storage service? It would require to wrap all storage method calls with these checks would be needed and throw a nicer exception? `ServiceLeftTheBuildingException` or something :grin:

Reply:
I'm not sure if this is an issue:

The `OAuthClientServiceImpl` instances hold their own reference of the OAuthStorageHandler. In case (with the changes of this PR in mind) the StorageService goes down, also the OAuthStorageHandler goes down and with it the OAuthFactory. Clients will not be able to create new OAuthClientServices any more. But the existing instances of `OAuthClientServiceImpl` still hold the reference to the old OAuthStorageHandler. There are some methods which do not depend on the internal StorageService (which is then null) but will continue to work, although from an OSGi pov the component is already deactivated.
The null checks which are now removed would anyway throw a runtime exception. I guess this was a little nicer then throwing NPE as it is now.

What the former dynamic dependency handling did support was that all depending services would _not_ go down when StorageService was deactivated. This we would lose now.

Reply:
In addition: When the `StorageService` decides to come back again the old version would renew the internal StorageFacade but it would still be the same reference of OAuthStorageService for all yet created OAuthClientService instances. With the new code those would be lost and the existing OAuthClientService instances would not benefit from the returning StorageService.

Reply:
On deactivating the `OAuthFactory` calls close on all created `OAuthClientService` by the factory. This close will set the storage handler to null in `OAuthClientService`. And it's can't be set again. So even when it would come back the `OAuthClientService` won't get it. To handle this a user of the service would always need to call `getOAuthClientService` on `OAuthFactory` when calling a method on the service and never keep a reference to `OAuthClientService` by itself. Or all service methods would add an extra parameter, the handle and be called via the factory. So the user never sees the `OAuthClientService`. Just some thoughts.

Reply:
After some more thinking. Is it correct that If the Storage will go down it will take OAuthFactory with it (also deactivate OAuthFactory) if storage is a required dependency? Does that then also mean the binding that is dependent on OAuthFactory will be deactivated because it has a required dependency?

Reply:
Ok slightly crossing posts :smile:
If what you suggest works we don't have to worry about the OAuthFactory. 
I'll try and see if it works.
Some additional questions:
1. Should the `unsetStorageService()` still call close on the `StorageFacade`?
1. I guess the null checks should be readded. I only thing they should end up with the user as `OAuthException`s and not as a `RuntimeException`. So should the StorageFacade throw these exceptions? or Should `OAuthServiceImpl` catch the `RuntimeException` and rethrow it as `OAuthException`?

Reply:
> Is it correct that If the Storage will go down it will take OAuthFactory with it (also deactivate OAuthFactory) if storage is a required dependency? 

Yes, if we make it a mandatory dependency.

> Does that then also mean the binding that is dependent on OAuthFactory will be deactivated because it has a required dependency?

Good point. Yes, if the binding handler factory depends on the OAuthFactory, it will actually mean that all handlers get disposed and the handler factory is deactivated BEFORE the OAuthFactory is deactivated and storages are removed. So unless the `OAuthClientService` instances are still used in asynchronous jobs (which should be all terminated by the handler upon disposal), we should be actually fine.